sycl/source/detail/context_impl.cpp

```diff
@@ -206,6 +225,58 @@ pi_native_handle context_impl::getNative() const {
   return Handle;
 }
 
+void context_impl::attachLifetimeToMemObj(std::shared_ptr<const void> &Resource,
+                                          const SYCLMemObjI *AttachTo) {
+  std::lock_guard<std::mutex> lock(MMemObjLifetimeAttachedResourcesMutex);
+  auto AttachedResourcesIt = MMemObjLifetimeAttachedResources.find(AttachTo);
+  if (AttachedResourcesIt != MMemObjLifetimeAttachedResources.end())
+    AttachedResourcesIt->second.push_back(Resource);
+  else
+    MMemObjLifetimeAttachedResources.insert({AttachTo, {Resource}});
+}
+
+void context_impl::detachMemObjLifetimeResources(
+    const SYCLMemObjI *AttachedTo) {
+  // Swap the attached resources and let them go out of scope without the lock.
+  // This is required as they could potentially have their own attached
+  // resources they need to detach.
+  std::vector<std::shared_ptr<const void>> AttachedResources;
+  {
+    std::lock_guard<std::mutex> lock(MMemObjLifetimeAttachedResourcesMutex);
+    auto AttachedResourcesIt =
+        MMemObjLifetimeAttachedResources.find(AttachedTo);
+    if (AttachedResourcesIt == MMemObjLifetimeAttachedResources.end())
+      return;
+    std::swap(AttachedResourcesIt->second, AttachedResources);
+    MMemObjLifetimeAttachedResources.erase(AttachedResourcesIt);
+  }
+}
+
+void context_impl::attachLifetimeToUSM(std::shared_ptr<const void> &Resource,
+                                       const void *AttachTo) {
+  std::lock_guard<std::mutex> lock(MUSMLifetimeAttachedResourcesMutex);
+  auto AttachedResourcesIt = MUSMLifetimeAttachedResources.find(AttachTo);
+  if (AttachedResourcesIt != MUSMLifetimeAttachedResources.end())
+    AttachedResourcesIt->second.push_back(Resource);
+  else
+    MUSMLifetimeAttachedResources.insert({AttachTo, {Resource}});
+}
+
+void context_impl::detachUSMLifetimeResources(const void *AttachedTo) {
+  // Swap the attached resources and let them go out of scope without the lock.
+  // This is required as they could potentially have their own attached
+  // resources they need to detach.
```

Comment:
Are  this comment and swapping relevant for USM pointers?

Reply:
Not at the moment as we don't have cases where an attached resource will be releasing a USM resource. That said, for future generality it will help prevent dead-locks if we ever have a reason to make "managed" USM attached resources.